### PR TITLE
Nvidia path for breadcrumbs

### PR DESCRIPTION
**Pull request by K0bin**
Based on #995

We could probably unify `vkd3d_breadcrumb_tracer_report_command_list_nv` and `vkd3d_breadcrumb_tracer_report_command_list_amd` but I didn't want to do that yet to keep the Nvidia path somewhat separate given that the breadcrumb PR is still marked as a draft.

### CODE REVIEW

File: libs/vkd3d/breadcrumbs.c [hunk 1]
```diff
@@ -167,6 +174,13 @@ void vkd3d_breadcrumb_tracer_cleanup(struct vkd3d_breadcrumb_tracer *tracer, str
         vkd3d_free_device_memory(device, &tracer->host_buffer_memory);
     }
 
+    for (i = 0; i < MAX_COMMAND_LISTS; i++) {
```
Comment (HansKristian-Work) [inline libs/vkd3d/breadcrumbs.c:177]: Formatting. Brace on next line.

File: libs/vkd3d/breadcrumbs.c [hunk 2]
```diff
@@ -356,9 +375,97 @@ static void vkd3d_breadcrumb_tracer_report_command_list_amd(struct vkd3d_breadcr
...
+static void vkd3d_breadcrumb_tracer_report_command_list_nv(struct vkd3d_breadcrumb_tracer *tracer,
+        unsigned int context_index, uint32_t begin_marker, uint32_t end_marker)
+{
+    const struct vkd3d_breadcrumb_command_list_trace_context *context;
+
+    context = &tracer->trace_contexts[context_index];
+
+    /* Unused, cannot be the cause. */
+    if (context->counter == 0)
+      return;
```
Comment (HansKristian-Work) [inline libs/vkd3d/breadcrumbs.c:387]: 4 spaces.

File: libs/vkd3d/breadcrumbs.c [hunk 3]
```diff
@@ -356,9 +375,97 @@ static void vkd3d_breadcrumb_tracer_report_command_list_amd(struct vkd3d_breadcr
...
+
+    ERR("Found pending command list context %u in executable state, TOP_OF_PIPE marker %u, BOTTOM_OF_PIPE marker %u.\n",
+            context_index, begin_marker, end_marker);
+    /* We set checkpoints for cmdlist begin, AFTER each command and cmdlist end.*/
+    vkd3d_breadcrumb_tracer_report_command_list(context, begin_marker, end_marker);
+    ERR("Done analyzing command list.\n");
+}
+
+static void vkd3d_breadcrumb_tracer_report_queue_nv(struct vkd3d_breadcrumb_tracer *tracer,
+                                                    struct d3d12_device *device,
```
Comment (HansKristian-Work) [inline libs/vkd3d/breadcrumbs.c:411]: Next line intent with 2x4 spaces.

File: libs/vkd3d/breadcrumbs.c [hunk 4]
```diff
@@ -356,9 +375,97 @@ static void vkd3d_breadcrumb_tracer_report_command_list_amd(struct vkd3d_breadcr
...
+    context_index = UINT32_MAX;
+    begin_marker = 0;
+    end_marker = 0;
+
+    for (i = 0; i < checkpoint_count; i++)
+    {
+        checkpoint = (struct vkd3d_nv_checkpoint*) checkpoints[i].pCheckpointMarker;
+        if (context_index != checkpoint->context_index && context_index != UINT32_MAX)
+        {
+            WARN("Markers have different contexts\n");
```
Comment (HansKristian-Work) [inline libs/vkd3d/breadcrumbs.c:441]: I assume this can happen if multiple queues are used concurrently?

File: libs/vkd3d/breadcrumbs.c [hunk 5]
```diff
@@ -468,6 +600,16 @@ void vkd3d_breadcrumb_tracer_signal(struct d3d12_command_list *list)
                         offsetof(struct vkd3d_breadcrumb_counter, begin_marker),
                 trace->counter + 1));
     }
+    else
+    {
+        checkpoint = vkd3d_malloc(sizeof(struct vkd3d_nv_checkpoint));
+        checkpoint->context_index = context;
+        checkpoint->counter = trace->counter;
+        VK_CALL(vkCmdSetCheckpointNV(list->vk_command_buffer, checkpoint));
```
Comment (HansKristian-Work) [inline libs/vkd3d/breadcrumbs.c:608]: It seems very pessimistic and slow to allocate a tiny cookie for each check point. Ideally, just encode the context index + breadcrumb index as a pointer directly. Even 32-bit pointer should be enough.

File: libs/vkd3d/breadcrumbs.c [hunk 6]
```diff
@@ -356,9 +375,97 @@ static void vkd3d_breadcrumb_tracer_report_command_list_amd(struct vkd3d_breadcr
...
+        checkpoint = (struct vkd3d_nv_checkpoint*) checkpoints[i].pCheckpointMarker;
+        if (context_index != checkpoint->context_index && context_index != UINT32_MAX)
+        {
+            WARN("Markers have different contexts\n");
+            context_index = UINT32_MAX;
+            break;
+        }
+        context_index = checkpoint->context_index;
+        if (checkpoints[i].stage == VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT)
+            begin_marker = checkpoint->counter;
```
Comment (HansKristian-Work) [inline libs/vkd3d/breadcrumbs.c:447]: Don't we want to look at first checkpoint for TOP_OF_PIPE, and last checkpoint for BOTTOM_OF_PIPE?

**Review by HansKristian-Work — CHANGES_REQUESTED** [review]

**Comment by K0bin** [comment]
Included in #995 now.

**Pull request closed without merging**